### ISSUE TITLE
0.13.3 on conda and fresh git-annex: duplicated output picked up within test_runner_parametrized_protocol?

### ISSUE THREAD

**Issue by yarikoptic** [datalad/datalad#4921]
@notestaff working on adding downstream testing back to git-annex package on conda-forge which is great!

datalad testing is not yet green.
[Full log](https://dev.azure.com/conda-forge/84710dde-1620-425b-80d0-4cf5baca359d/_apis/build/builds/209014/logs/13) (160MB huge  due to annexremote testing verbosity, unrelated), but then testing of datalad fails with a single (better than if many) test failing

```
2020-09-14T06:57:15.1594851Z Versions: annexremote=1.4.3 appdirs=1.4.3 boto=2.49.0 cmd:7z=16.02 cmd:annex=8.20200908-ga853039 cmd:bundled-git=2.28.0 cmd:git=2.28.0 cmd:system-git=2.28.0 cmd:system-ssh=8.3p1 exifread=2.1.2 humanize=2.6.0 iso8601=0.1.13 keyring=21.4.0 keyrings.alt=3.1 msgpack=1.0.0 mutagen=1.41.1 patoolib=1.12 requests=2.24.0 tqdm=4.49.0 wrapt=1.12.1
2020-09-14T06:57:15.1601769Z Obscure filename: str=b' "\';a&b&c\xce\x94\xd0\x99\xd7\xa7\xd9\x85\xe0\xb9\x97\xe3\x81\x82 `| ' repr=' "\';a&b&cΔЙקﻡ๗あ `| '
2020-09-14T06:57:15.1608345Z Encodings: default='utf-8' filesystem='utf-8' locale.prefered='UTF-8'
2020-09-14T06:57:15.1616112Z Environment: PYTHONNOUSERSITE='1' PYTHONUNBUFFERED='1' PATH='$PREFIX:$PREFIX/bin:$PREFIX/bin:/opt/conda/condabin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/home/conda/bin' LANG='en_US.UTF-8' GIT_BRANCH='merge' PYTHON='$PREFIX/bin/python' GIT_CONFIG_PARAMETERS="'init.defaultBranch=dl-test-branch'"
2020-09-14T06:57:15.2906560Z Exception ignored in: <function _Database.close at 0x7fa3c4c100d0>
2020-09-14T06:57:15.2907546Z Traceback (most recent call last):
2020-09-14T06:57:15.2909376Z   File "/home/conda/feedstock_root/build_artifacts/git-annex_1600056447586/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/python3.8/dbm/dumb.py", line 274, in close
2020-09-14T06:57:15.2910430Z     self._commit()
2020-09-14T06:57:15.2911738Z   File "/home/conda/feedstock_root/build_artifacts/git-annex_1600056447586/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/python3.8/dbm/dumb.py", line 129, in _commit
2020-09-14T06:57:15.2913031Z     with self._io.open(self._dirfile, 'w', encoding="Latin-1") as f:
2020-09-14T06:57:15.2914219Z FileNotFoundError: [Errno 2] No such file or directory: '/tmp/datalad_temp_y1aehq1r/.config/datalad/cookies.dir'
2020-09-14T06:57:15.2914736Z
2020-09-14T06:57:15.2915093Z ======================================================================
2020-09-14T06:57:15.2915566Z FAIL: datalad.tests.test_witless_runner.test_runner_parametrized_protocol
2020-09-14T06:57:15.2916282Z ----------------------------------------------------------------------
2020-09-14T06:57:15.2916765Z Traceback (most recent call last):
2020-09-14T06:57:15.2918077Z   File "/home/conda/feedstock_root/build_artifacts/git-annex_1600056447586/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/python3.8/site-packages/nose/case.py", line 197, in runTest
2020-09-14T06:57:15.2919171Z     self.test(*self.arg)
2020-09-14T06:57:15.2920772Z   File "/home/conda/feedstock_root/build_artifacts/git-annex_1600056447586/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/python3.8/site-packages/datalad/tests/test_witless_runner.py", line 149, in test_runner_parametrized_protocol
2020-09-14T06:57:15.2922014Z     eq_(res['stdout'], '5')
2020-09-14T06:57:15.2922684Z AssertionError: '55' != '5'
2020-09-14T06:57:15.2923417Z -------------------- >> begin captured logging << --------------------
2020-09-14T06:57:15.2924415Z asyncio: DEBUG: Using selector: EpollSelector
2020-09-14T06:57:15.2925346Z --------------------- >> end captured logging << ---------------------
2020-09-14T06:57:15.2925748Z
2020-09-14T06:57:15.2926393Z ----------------------------------------------------------------------
2020-09-14T06:57:15.2926856Z Ran 1163 tests in 1886.415s
2020-09-14T06:57:15.2927141Z
2020-09-14T06:57:15.2927419Z FAILED (SKIP=122, failures=1)
```

**Comment by kyleam** [comment]
Here's my vague guess at what's happening.

```python
    runner = Runner()

    # protocol returns a given value whatever it receives
    class ProtocolInt(StdOutCapture):
        def __init__(self, done_future, value):
            self.value = value
            super().__init__(done_future)

        def pipe_data_received(self, fd, data):
            super().pipe_data_received(fd, self.value)

    res = runner.run(
        py2cmd('print(1)'),
        protocol=ProtocolInt,
        # value passed to protocol constructor
        value=b'5',
    )
    eq_(res['stdout'], '5')
```

So, to check that the keyword argument passed to `run` is relayed to the protocol, this test replaces the incoming data with the value each time `pipe_data_received` is called.  The assumption is that `print(1)` causes `pipe_data_received` to be called just once.  That value would be duplicated if `pipe_data_received` receives two chunks of stdout (1 and then new line? other stdout?).

If that is the cause, we could work around it with something like this:

```diff
diff --git a/datalad/tests/test_witless_runner.py b/datalad/tests/test_witless_runner.py
index 08c60f94b..86f4c2ecc 100644
--- a/datalad/tests/test_witless_runner.py
+++ b/datalad/tests/test_witless_runner.py
@@ -135,10 +135,12 @@ def test_runner_parametrized_protocol():
     class ProtocolInt(StdOutCapture):
         def __init__(self, done_future, value):
             self.value = value
+            self.pipe_data_called = False
             super().__init__(done_future)
 
         def pipe_data_received(self, fd, data):
-            super().pipe_data_received(fd, self.value)
+            if not self.pipe_data_called:
+                super().pipe_data_received(fd, self.value)
 
     res = runner.run(
         py2cmd('print(1)'),
```

**Comment by yarikoptic** [comment]
Thank you @kyleam for looking into it.  I have not tried to reproduce locally yet but quickly initiated a new build on conda-forge: https://github.com/conda-forge/datalad-feedstock/compare/master...yarikoptic:test-build?expand=1 -- let's see if it fails and then we could even try any perspective patch there

**Comment by kyleam** [comment]
> I have not tried to reproduce locally yet

I've tried in a loop with no luck, which isn't surprising given that, as far as I know, we haven't seen this failure before.

**Comment by yarikoptic** [comment]
yeap, I failed to replicate locally as well so far :-/

**Comment by mih** [comment]
This was reported on a version that is rather old now. Still relevant?

**Comment by yarikoptic** [comment]
talks about witless (async) runner which is no longer.  downstream testing is on and off from time to time but I do not remember seeing this particular issue for a while, and we do not exclude this test. Let's consider it resolved by switching to non-async runnger by @christian-monch

**Comment by yarikoptic** [comment]
nope -- unfortunately it is rare but still not gone: https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=489377&view=logs&j=4f922444-fdfe-5dcf-b824-02f86439ef14&t=b2a8456a-fb11-5506-ca32-5ccd32538dc0 while building 0.16.1 against python 3.8
```
======================================================================
FAIL: datalad.runner.tests.test_witless_runner.test_runner_parametrized_protocol
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/conda/feedstock_root/build_artifacts/datalad_1649524824072/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/lib/python3.8/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/conda/feedstock_root/build_artifacts/datalad_1649524824072/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/lib/python3.8/site-packages/datalad/runner/tests/test_witless_runner.py", line 202, in test_runner_parametrized_protocol
    eq_(res['stdout'], '5')
AssertionError: '55' != '5'
```

may be @christian-monch  has an idea?  (meanwhile I will force push rewrite to trigger a rerun of CI)

**Comment by yarikoptic** [comment]
ping @christian-monch -- happened again in https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=494689&view=logs&j=4f922444-fdfe-5dcf-b824-02f86439ef14&t=b2a8456a-fb11-5506-ca32-5ccd32538dc0 and again for 3.8 on linux, so consistent in that!  I will restart the build

**Comment by yarikoptic** [comment]
happened again for 0.17.1 pkg build and again on python 3.8 only https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=534639&view=logs&j=4f922444-fdfe-5dcf-b824-02f86439ef14&t=b2a8456a-fb11-5506-ca32-5ccd32538dc0  .  I think we should just announce that test `xfail` for the python 3.8 and continue with our lives since smells like some bug in python which was addressed after. WDYT?

**Comment by yarikoptic** [comment]
it was closed by mistake -- not yet resolved. And now happened for 3.9 as well: https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=556819&view=logs&j=517fe804-fa30-5dc2-1413-330699242c05&t=c10fa5f2-fdf6-5338-3bdb-c4bea7c23412 while building for conda.

**Comment by christian-monch** [comment]
I suspect the problem is that the printed string has two characters, ie. `"1"` and `"\n"`. Depending on the system state the pipe might signal `"1\n"`, or it might signal `"1"` and `"\n"`. In the latter case, `self.values`, i.e. `5`, would be appended two times to `stdout`.

**Comment by yarikoptic** [comment]
Thanks @christian-monch for the analysis!!! I fail to see how/why changing the test "fixes" the situation in https://github.com/christian-monch/datalad/commit/bc4ded8c88d80089d2e0457f01bbb3f6f2457bab#diff-472642261afcc22bd6f543b4a8c20533588ec9281bd964137b29f79885e56164L197 ?  Test code seems to be totally legit - just running a command which prints a single character, so should not be changed.  IMHO it is the code of the runner is wrong somewhere that it manages to duplicate that single character, so better be fixed in the code (not in the test). Or what am I missing?

**Comment by christian-monch** [comment]
Should be fixed by PR #6978. Closing this issue
